File: ios/Wavepoint/Auth/SupabaseSpotifyAuthenticator.swift
import AuthenticationServices
import Foundation
import Supabase

struct SpotifyAuthorizationRequest: Equatable, Sendable {
  let provider: Provider
  let callbackURL: URL
  let scopes: String

  var callbackScheme: String {
    callbackURL.scheme ?? ""
  }

  init(callbackURL: URL) throws {
    guard callbackURL.scheme != nil else {
      throw SpotifyAuthorizationRequestError.missingCallbackScheme
    }

    provider = .spotify
    self.callbackURL = callbackURL
    scopes = [
      "user-read-email",
      "user-read-private",
      "user-library-read",
      "user-library-modify",
      "user-read-recently-played",
      "app-remote-control",
    ].joined(separator: " ")
  }
}

enum SpotifyAuthorizationRequestError: Error, Equatable {
  case missingCallbackScheme
}

struct SupabaseSpotifyAuthenticator: SpotifyAuthenticating {
  private let client: SupabaseClient
  private let request: SpotifyAuthorizationRequest

  init(client: SupabaseClient, callbackURL: URL) throws {
    self.client = client
    request = try SpotifyAuthorizationRequest(callbackURL: callbackURL)
  }

  static func makeClient(configuration: AppConfiguration) -> SupabaseClient {
    SupabaseClient(
      supabaseURL: configuration.supabaseURL,
      supabaseKey: configuration.supabasePublishableKey,
      options: SupabaseClientOptions(
        auth: .init(
          redirectToURL: configuration.callbackURL,
          flowType: .pkce
        )
      )
    )
  }

  func restoreSession() async throws -> SpotifyAuthSession? {
    guard client.auth.currentSession != nil else { return nil }
    return SpotifyAuthSession(session: try await client.auth.session)
  }

  func signIn() async throws -> SpotifyAuthSession {
    let session = try await client.auth.signInWithOAuth(
      provider: request.provider,
      redirectTo: request.callbackURL,
      scopes: request.scopes
    ) { session in
      session.prefersEphemeralWebBrowserSession = false
    }
    return SpotifyAuthSession(session: session)
  }

  func signOut() async throws {
    try await client.auth.signOut()
  }

  func clearLocalSession() async throws {
    try await client.auth.signOut(scope: .local)
  }
}

extension SpotifyAuthSession {
  fileprivate init(session: Session) {
    let tokens = session.providerToken.map {
      SpotifyProviderTokens(
        accessToken: $0,
        refreshToken: session.providerRefreshToken
      )
    }
    self.init(providerTokens: tokens)
  }
}

File: android/app/src/main/java/ai/mapier/swipe/auth/SupabaseSpotifyAuthenticator.kt
package ai.mapier.swipe.auth

import android.content.Intent
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.Auth
import io.github.jan.supabase.auth.ExternalAuthAction
import io.github.jan.supabase.auth.FlowType
import io.github.jan.supabase.auth.SignOutScope
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.handleDeeplinks
import io.github.jan.supabase.auth.providers.Spotify
import io.github.jan.supabase.auth.user.UserSession
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.functions.Functions

class SupabaseSpotifyAuthenticator(
  private val client: SupabaseClient,
  private val request: SpotifyAuthorizationRequest = SpotifyAuthorizationRequest(),
) : SpotifyAuthenticator {
  override suspend fun restoreSession(): SpotifyAuthSession? {
    client.auth.awaitInitialization()
    return client.auth.currentSessionOrNull()?.toDomain()
  }

  override suspend fun startSignIn() {
    client.auth.signInWith(Spotify) {
      scopes.addAll(request.scopes)
    }
  }

  fun handleDeepLink(
    intent: Intent,
    onSession: (SpotifyAuthSession) -> Unit,
    onError: (Throwable) -> Unit,
  ) {
    client.handleDeeplinks(
      intent = intent,
      onSessionSuccess = { onSession(it.toDomain()) },
      onError = onError,
    )
  }

  override suspend fun signOut() {
    client.auth.signOut()
  }

  override suspend fun clearLocalSession() {
    client.auth.signOut(SignOutScope.LOCAL)
  }

  private fun UserSession.toDomain() = SpotifyAuthSession(
    supabaseAccessToken = accessToken,
    providerTokens = providerToken?.let { accessToken ->
      SpotifyProviderTokens(
        accessToken = accessToken,
        refreshToken = providerRefreshToken,
      )
    },
  )

  companion object {
    fun createClient(
      supabaseUrl: String,
      publishableKey: String,
      request: SpotifyAuthorizationRequest = SpotifyAuthorizationRequest(),
    ): SupabaseClient = createSupabaseClient(supabaseUrl, publishableKey) {
      install(Auth) {
        flowType = FlowType.PKCE
        scheme = request.callbackScheme
        host = request.callbackHost
        defaultExternalAuthAction = ExternalAuthAction.CustomTabs()
      }
      install(Functions)
    }
  }
}
